File: internal/sniproxy/http.go
// Copyright (C) 2019-2022 Antoine Tenart <antoine.tenart@ack.tf>
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program. If not, see <https://www.gnu.org/licenses/>.

package sniproxy

import (
	"bufio"
	"bytes"
	"fmt"
	"io"
	"net"
	"net/http"

	"github.com/atenart/sniproxy/internal/log"
)

// Handles HTTP requests by starting an HTTP server answering 308 redirects if
// the requested hostname is one we serve and if the redirection is allowed in
// the configuration for said hostname.
//
// Before calling this a check using Config.NeedHTTP() first can be made to
// ensure an HTTP server is needed, depending on the provided configuration.
func (p *Proxy)HandleRedirect(bind string) error {
	return http.ListenAndServe(bind, http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		host := cleanHost(r.Host)
		redirHost, status, err := getRedirectHost(&p.Config, host, REDIRECT_FROM_HTTP)
		if err != nil {
			log.Printf(log.INFO, "%s (%s) - %s", r.RemoteAddr, host, err)
			http.Error(w, http.StatusText(status), status)
			return
		}

		url := fmt.Sprintf("https://%s%s", redirHost, r.RequestURI)
		http.Redirect(w, r, url, status)

		log.Printf(log.INFO, "%s (%s) - Redirecting request to HTTPS", r.RemoteAddr, host)
	}))
}

// Checks if the data in buf looks like an HTTP request. This does not guarantee
// the request is one, but should be enough to at least try handling it.
func isHTTP(buf *bytes.Buffer) bool {
	// The request is first parsed by parseRecord() which will try to read
	// a record header first. The record header is 5 bytes long.
	//
	// In case parseRecord() fails to read the record header, the buffer
	// won't contain the first 5 bytes. This is a request issue, we have no
	// reason to retry reading here.
	//
	// In all other cases at a minimum 5 bytes will be in the buffer, so
	// match on that. This is convenient as matching on more would require
	// some extra logic (method names do not all have the same length).
	// (We're only trying to loosely identify an HTTP request).
	if buf.Len() >= 5 {
		switch buf.String()[:5] {
		// From https://developer.mozilla.org/fr/docs/Web/HTTP/Methods
		case "GET /", "HEAD ", "POST ", "PUT /", "DELET", "CONNE", "OPTIO", "TRACE", "PATCH":
			return true
		}
	}
	return false
}

// Try redirecting what could be an HTTP request received on a TCP socket (in
// our case the TLS SNI server).
func redirectHTTP(conn *Conn, buf *bytes.Buffer) {
	req, err := http.ReadRequest(bufio.NewReader(io.MultiReader(buf, conn)))
	if err != nil {
		// Log as DEBUG, remember this is best effort and we're not sure
		// the request is genuine.
		conn.logf(log.DEBUG, "Could not parse the HTTP request: %s", err)
		return
	}

	host := cleanHost(req.Host)
	conn.sni = host
	redirHost, status, err := getRedirectHost(conn.Config, host, REDIRECT_FROM_TLS)
	if err != nil {
		conn.logf(log.INFO, "HTTP request on TLS port: %s", err)
	}

	hdr := make(http.Header)
	if status == http.StatusPermanentRedirect {
		hdr.Add("Location", fmt.Sprintf("https://%s%s", redirHost, req.RequestURI))
	}

	response := &http.Response{
		Proto: "HTTP/1.0",
		ProtoMajor: 1,
		ProtoMinor: 0,
		StatusCode: status,
		Status: http.StatusText(status),
		Header: hdr,
	}
	if err := response.Write(conn); err != nil {
		conn.logf(log.ERR, "Could not send the HTTP response: %s", err)
		return
	}

	if status == http.StatusPermanentRedirect {
		conn.logf(log.INFO, "Redirecting request to HTTPS")
	}
}

func getRedirectHost(c *Config, host string, redirectType uint32) (string, int, error) {
	route := c.MatchBackend(host)
	if route == nil {
		err := fmt.Errorf("No route to domain %s", host)
		return "", http.StatusForbidden, err
	} else if route.HTTPRedirect == REDIRECT_NONE || (route.HTTPRedirect & redirectType) == 0 {
		err := fmt.Errorf("Request to %s not being redirected", host)
		return "", http.StatusForbidden, err
	}

	if c.RedirectPort != 0 && c.RedirectPort != 443 {
		host = fmt.Sprintf("%s:%d", host, c.RedirectPort)
	}

	return host, http.StatusPermanentRedirect, nil
}

func cleanHost(from string) string {
	host, _, err := net.SplitHostPort(from)
	if err != nil {
		// No port in reqHost.
		host = from
	}

	return host
}

File: src/http.rs
use std::{
    io::{Read, Write},
    net::SocketAddr,
    str,
};

use anyhow::{Result, bail};
use log::info;

use crate::{config::Config, reader::ReaderBuf};

/// Check if the provided buffer looks like an HTTP request. This does not guarantee the request is
/// a genuine one, but should be enough to at least try handling it.
pub(crate) fn is_http<R: Read>(rb: &ReaderBuf<R>) -> bool {
    // The buffer comes from previous reads, we might have failed to read up to 5 bytes. No reason
    // try try reading it again. In all other cases 5 bytes is the maximal length we can match all
    // HTTP methods on. This is a shortcut but as we don't try to be 100% correct here, this is
    // fine.
    let buf = rb.buf();
    if buf.len() >= 5 {
        // From https://developer.mozilla.org/fr/docs/Web/HTTP/Methods
        if let Ok(
            "GET /" | "HEAD " | "POST " | "PUT /" | "DELET" | "CONNE" | "OPTIO" | "TRACE" | "PATCH",
        ) = str::from_utf8(&buf[..5])
        {
            return true;
        }
    }
    false
}

/// Redirect an HTTP request with a 308.
pub(crate) fn try_redirect<R: Read + Write>(
    config: &Config,
    client: &SocketAddr,
    rb: &mut ReaderBuf<R>,
) -> Result<()> {
    let host = match get_host(rb)? {
        Some(host) => host,
        None => return Ok(()), // Not much we can do, not really an error.
    };
    #[cfg(not(test))]
    crate::context::set_hostname(host)?;

    match config.get_route(host) {
        Some(route) => {
            if !route.http_redirect {
                bail!("Denied HTTP redirect to HTTPS (disabled)");
            }

            if !route.is_allowed(client) {
                bail!("Denied HTTP redirect to HTTPS (ACLs)");
            }
        }
        None => bail!("Unknown hostname ({host})"), // Not much we can do, not really an error.
    }

    info!("Redirecting HTTP request to HTTPS");
    let response = format!(
        "HTTP/1.0 308 Unknown\r\nLocation: https://{host}:{}\r\n\r\n",
        config.bind_https.port()
    );
    Ok(rb.get_mut().write_all(response.as_bytes())?)
}

/// Try parsing an HTTP request host.
fn get_host<R: Read>(rb: &mut ReaderBuf<R>) -> Result<Option<&str>> {
    // Try to read the remaining of the HTTP headers. 8KB is the limit size on
    // many web servers.
    rb.read(8192)?;
    let headers = str::from_utf8(rb.buf())?;

    // Skip the request line and loop over the HTTP headers to find the Host
    // one and extract its value.
    for hdr in headers.split("\r\n").skip(1) {
        match hdr.split_once(':') {
            Some(("Host", val)) => {
                let val = val.trim();
                // Host could be in the <host>:<port> form.
                return Ok(Some(match val.split_once(':') {
                    Some((host, _)) => host,
                    _ => val,
                }));
            }
            _ => continue,
        }
    }

    info!("Could not find HTTP Host header in request");
    Ok(None)
}

#[cfg(test)]
mod tests {
    use std::io::Cursor;

    use crate::{config::Config, reader::ReaderBuf as B, tls::tests::RECORD_SNI_ALPN};

    #[test]
    fn is_http() {
        let valid: &[&str] = &[
            "GET /", "HEAD ", "POST ", "PUT /", "DELET", "CONNE", "OPTIO", "TRACE", "PATCH",
        ];
        for s in valid.iter() {
            let mut rb = &mut B::from_bytes(s.as_bytes());
            rb.read(5).unwrap();
            assert!(super::is_http(&mut rb));
        }

        let mut rb = B::from_bytes(&[]);
        rb.read(5).unwrap();
        assert_eq!(super::is_http(&mut rb), false);

        let mut rb = B::from_bytes(&[0; 256]);
        rb.read(5).unwrap();
        assert_eq!(super::is_http(&mut rb), false);
    }

    #[test]
    fn get_host() {
        assert_eq!(super::get_host(&mut B::from_bytes(&[])).unwrap(), None);
        assert_eq!(super::get_host(&mut B::from_bytes(b"GET /")).unwrap(), None);
        assert_eq!(
            super::get_host(&mut B::from_bytes(b"GET /\r\nHost: example.net\r\n")).unwrap(),
            Some("example.net")
        );
        assert_eq!(
            super::get_host(&mut B::from_bytes(
                b"GET /\r\nHost: foo.example.net:8080\r\n"
            ))
            .unwrap(),
            Some("foo.example.net")
        );
        assert_eq!(
            super::get_host(&mut B::from_bytes(
                b"GET /\r\nScheme: https\r\nHost: example.net\r\nFilename: foo\r\n"
            ))
            .unwrap(),
            Some("example.net")
        );

        // Invalid UTF-8 sequence.
        assert!(super::get_host(&mut B::from_bytes(RECORD_SNI_ALPN)).is_err());
    }

    #[test]
    fn try_redirect() {
        let config = Config::from_str(
            "
routes:
  - domains:
    - example.net
    backend:
      address: 127.0.0.1:8000
  - domains:
    - denied.example.net
    http_redirect: false
    backend:
      address: 127.0.0.1:8000
  - domains:
    - acls.example.net
    backend:
      address: 127.0.0.1:8000
    allowed_ranges:
    - 10.0.0.0/8
    denied_ranges:
    - 127.0.0.0/24
        ",
        )
        .unwrap();

        let req = "GET /\r\nHost: example.net\r\n".as_bytes();
        let mut rb = B::new(Cursor::new(req.to_vec()));
        assert!(super::try_redirect(&config, &"127.0.0.1:10000".parse().unwrap(), &mut rb).is_ok());

        let buf = rb.into_inner().into_inner();
        assert_eq!(
            &buf[req.len()..],
            b"HTTP/1.0 308 Unknown\r\nLocation: https://example.net:443\r\n\r\n"
        );

        let req = "GET /\r\n".as_bytes();
        let mut rb = B::new(Cursor::new(req.to_vec()));
        assert!(super::try_redirect(&config, &"127.0.0.1:10000".parse().unwrap(), &mut rb).is_ok());

        let req = "GET /\r\nHost: denied.example.net\r\n".as_bytes();
        let mut rb = B::new(Cursor::new(req.to_vec()));
        assert!(
            super::try_redirect(&config, &"127.0.0.1:10000".parse().unwrap(), &mut rb).is_err()
        );

        let req = "GET /\r\nHost: foo.example.net\r\n".as_bytes();
        let mut rb = B::new(Cursor::new(req.to_vec()));
        assert!(
            super::try_redirect(&config, &"127.0.0.1:10000".parse().unwrap(), &mut rb).is_err()
        );

        let req = "GET /\r\nHost: acls.example.net\r\n".as_bytes();
        let mut rb = B::new(Cursor::new(req.to_vec()));
        assert!(
            super::try_redirect(&config, &"127.0.0.1:10000".parse().unwrap(), &mut rb).is_err()
        );

        let req = "GET /\r\nHost: acls.example.net\r\n".as_bytes();
        let mut rb = B::new(Cursor::new(req.to_vec()));
        assert!(super::try_redirect(&config, &"10.0.42.1:10000".parse().unwrap(), &mut rb).is_ok());
    }
}
